In [18]:
class CountReason:
    def __init__(self, p_c, p_l, p_lc):
        self.p_c = p_c
        self.p_l = p_l
        self.p_lc = p_lc
        self.p_cl = None

    def count_reason(self):
        self.p_cl = (self.p_lc * self.p_c) / self.p_l

### Przypadek #1:

Wysokie prawdopodobieństwo utraty węchu, pod warunkiem zachorowania na COVID

In [19]:
p_c = 0.0062
p_l = 0.01
p_lc = 0.9

covid = CountReason(p_c, p_l, p_lc)
covid.count_reason()
print(f"Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:\n{covid.p_cl * 100:.2f}%")

Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:
55.80%


### Przypadek #2:

Utrata węchu jest bardziej popularnym zjawiskiem

In [20]:
p_c = 0.0062
p_l = 0.1
p_lc = 0.9

covid = CountReason(p_c, p_l, p_lc)
covid.count_reason()
print(f"Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:\n{covid.p_cl * 100:.2f}%")

Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:
5.58%


### Przypadek #3:

Mniejsze prawdodobieństwo występowania utraty węchu podczas zakarzenia COVIDem

In [21]:
p_c = 0.0062
p_l = 0.01
p_lc = 0.1

covid = CountReason(p_c, p_l, p_lc)
covid.count_reason()
print(f"Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:\n{covid.p_cl * 100:.2f}%")

Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:
6.20%


### Przypadek #4:

Obliczanie na podstawie BRAKU utraty węchu (czyli szansa na zachorowanie bez objawów).

In [22]:
p_c = 0.0062
p_l = 0.01
p_not_l = 1 - p_l
p_lc = 0.9
p_not_lc = 1 - p_lc

covid = CountReason(p_c, p_not_l, p_not_lc)
covid.count_reason()
print(f"Prawdopodobieństwo zachorowania na COVID pod warunkiem braku utraty węchu:\n{covid.p_cl * 100:.2f}%")

Prawdopodobieństwo zachorowania na COVID pod warunkiem braku utraty węchu:
0.06%


### Przypadek #5:

Spora szansa na zachorowanie na COVID, bardzo małe szanse utraty węchu, rzadka utrata węchu pod warunkiem zachorowania na COVID.

In [26]:
p_c = 0.2
p_l = 0.0001
p_lc = 0.01

covid = CountReason(p_c, p_l, p_lc)
covid.count_reason()
print(f"Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:\n{covid.p_cl * 100:.2f}%")

Prawdopodobieństwo zachorowania na COVID pod warunkiem utraty węchu:
2000.00%


Prawdopodobieństwo wyszło nieprawidłowe, głównie przez bardzo niską szansę na utratę węchu. 

Obliczając też dodatkowo prawdopodobieństwo utraty węchu mając COVID:

In [27]:
expected_loss_from_covid = p_c * p_lc
print(f"P-stwo utraty węchu przez COVID: {expected_loss_from_covid*100:.2f}%")

P-stwo utraty węchu przez COVID: 0.20%


Prawdopodobieństwo utraty węchu PRZEZ COVIDa wychodzi większe niż całkowita szansa na utratę węchu: 

**0.2% > 0.1%**
co jest nierealne.

In [28]:
print("P(L) = P(L|C) * P(C) + P(L|not C) * P(not C)")

p_l_not_c = 0.0001  
p_not_c = 1 - p_c

p_l_new = p_lc * p_c + p_l_not_c * p_not_c
print(f"P(L|not C) (utrata węchu pod warunkiem braku COVID) = {p_l_not_c}")
print(f"P(not C) = {p_not_c}")
print(f"P(L) = {p_lc} * {p_c} + {p_l_not_c} * {p_not_c}")
print(f"P(L) = {p_l_new*100:.2f}%")

covid = CountReason(p_c, p_l_new, p_lc)
covid.count_reason()
print(f"Nowy wynik z użyciem prawa całkowitego prawdopodobieństwa: {covid.p_cl*100:.2f}%")

P(L) = P(L|C) * P(C) + P(L|not C) * P(not C)
P(L|not C) (utrata węchu pod warunkiem braku COVID) = 0.0001
P(not C) = 0.8
P(L) = 0.01 * 0.2 + 0.0001 * 0.8
P(L) = 0.21%
Nowy wynik z użyciem prawa całkowitego prawdopodobieństwa: 96.15%


Zakładając prawdopodobieństwo utraty węchu pod warunkiem BRAKU zachorowania na COVID, mogliśmy użyć prawa całkowitego prawdopodobieństwa i ze wzoru obliczyć nowe prawdopodobieństwo utraty węchu. Wynosi ono teraz 0.21% (więcej niż 0.01%).

Teraz wynik jest prawidłowy (założyliśmy bardziej realne dane) - znajduje się w przedziale 0-1.